packages/type-utils/src/isTypeReadonly.ts

```diff
@@ -31,23 +36,65 @@ export const readonlynessOptionsSchema = {
     treatMethodsAsReadonly: {
       type: 'boolean',
     },
+    allowlist: {
+      type: 'array',
+      items: typeAllowListItemSchema,
+    },
   },
 };
 
 export const readonlynessOptionsDefaults: ReadonlynessOptions = {
   treatMethodsAsReadonly: false,
+  allowlist: [],
 };
 
 function hasSymbol(node: ts.Node): node is ts.Node & { symbol: ts.Symbol } {
   return Object.prototype.hasOwnProperty.call(node, 'symbol');
 }
 
+function isTypeExcepted(
+  type: ts.Type,
+  program: ts.Program,
+  options: ReadonlynessOptions,
+): boolean {
+  const typeName = type.getSymbol()?.escapedName;
+  const matchingItems =
+    options.allowlist?.filter(item => item.typeName === typeName) ?? [];
+  for (const item of matchingItems) {
+    const declarationFiles =
+      type
+        .getSymbol()
+        ?.getDeclarations()
+        ?.map(declaration => declaration.getSourceFile()) ?? [];
+    for (const declaration of declarationFiles) {
+      if (
+        // A local type defined in the current package
+        (item.source === 'local' &&
+          declaration.fileName.startsWith(program.getCurrentDirectory())) ||
+        // A type from the default library
+        (item.source === 'default-lib' &&
+          program.isSourceFileDefaultLibrary(declaration)) ||
+        // A type from a specified third-party package
+        (item.source === 'package' &&
+          (declaration.fileName.includes('node_modules/' + item.package) ||
+            declaration.fileName.includes(
+              'node_modules/@types/' + item.package,
+            )))
+      ) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
 function isTypeReadonlyArrayOrTuple(
-  checker: ts.TypeChecker,
+  program: ts.Program,
```

Comment:
unfortunately given that this function forms part of this package's public API - this parameter change is a breaking change.

we will need to either make this change in a backwards compatible way, or this change will need to wait for the next major.

Reply:
Hi,
that's too bad - I don't see a way of doing this without a `Program` and I don't think you can get the program from the checker :(. So if we were to do this in a backwards-compatible way, we'd have to add the program as another optional parameter and gate the whitelist functionality with it - a very messy solution. Is there any ETA on a next major version? It seems to me that the `type-utils` package isn't used much outside of `typescript-eslint`, so a major version isn't a big deal, right?

Reply:
> Is there any ETA on a next major version?

No - we don't really "plan" these things. We do them when there's a good motivating case to go through the whole rigamarole. It's a fair chunk of effort to coordinate the branches, tagged releases and such to prepare a major - so we want a really good reason to go through the pain!

> It seems to me that the type-utils package isn't used much outside of typescript-eslint, so a major version isn't a big deal, right?

We have a single version for all of our packages. This means we can't bump just one - we have to bump them all together.
This is done because it makes it simpler for consumers to understand what to do.
I.e. `vX.Y.Z` of package A should be used with `vX.Y.Z` of package B.

Reply:
Ok, well then this PR will have to wait for the next major :(

Reply:
On the bright side, we're now actively merging things into v6! 🎉 😄